klippy/extras/mixing_extruder.py

```diff
@@ -0,0 +1,335 @@
+        gcode_pos = self.gcode_move.get_status()['gcode_position']
+        self._apply_mixing(self._get_gradient(gcode_pos))
+    def _disable_gradient(self):
+        if not self.normal_transform:
+            return
+        self.gcode_move.set_move_transform(self.normal_transform, force=True)
+        self.normal_transform = None
+    def _get_gradient(self, pos):
+        mixing = self._mixing
+        for heights, refs in self._gradients:
+            start, _, end = heights
+            start_mix, end_mix = refs
+            if self._gradient_method == 'linear':
+                zpos = sum(pos[i]*self._gradient_vector[i] for i in range(3))
+                if zpos <= start:
+                    return start_mix
+                if zpos >= end:
+                    mixing = end_mix
+                    continue
+                w = (zpos - start) / (end - start)
+                return list(((1. - w) * s + w * e)
+                            for s, e in zip(start_mix, end_mix))
+            if self._gradient_method == 'spherical':
+                dist = math.sqrt(sum((x-self._gradient_vector[i])**2
+                                     for i, x in enumerate(pos)))
+                if dist <= start:
+                    return start_mix
+                if dist >= end:
+                    mixing = end_mix
+                    continue
+                w = (dist - start) / (end - start)
+                mix = list(((1. - w) * s + w * e)
+                           for s, e in zip(start_mix, end_mix))
+                return mix
+        return mixing
+    def get_position(self):
+        self.normal_transform.get_position()
+    def move(self, newpos, speed):
+        gcode_pos = self.gcode_move.get_status()['gcode_position']
+        self._apply_mixing(self._get_gradient(gcode_pos))
+        self.normal_transform.move(newpos, speed)
+    def get_status(self, eventtime):
+        status = super(GradientMixingExtruder, self).get_status(eventtime)
+        for i, gradient in enumerate(self._gradients):
+            status.update({"mixing_gradient%d" % (i): ",".join(
+                "%s:%s" % (k, v)
+                for k, v in dict(
+                    heights="%.1f-(%.1f)-%.1f" % gradient[0],
+                    mixings="%s-%s" % tuple(
+                        "/".join("%.1f" % (x) for x in m)
+                        for m in gradient[1]),
+                    method=self._gradient_method,
+                    vector="/".join("%.1f" % (x)
+                                    for x in self._gradient_vector),
+                    enabled=str(self._gradient_enabled)).items())})
+        return status
+    def _check_scales(self, gcmd, s):
+        try:
+            scales = s.replace(',', ':').split(':')
+            if len(scales) != len(self._mixing):
+                raise gcmd.error("Could not configure gradient: invalid scales")
+            return tuple(float(scales[i]) for i in range(len(scales)))
+        except Exception:
```

Comment:
It's not a good idea to catch any Exception as that often results in real code bugs not being reported properly.  (A real bug will end up corrupting internal state and when the code finally does error, the reported error wont indicate the real source of the issue.)  Prefer to catch just the exception needed.

Reply:
Have adjusted it to ValueException which may originated from a non-number passed from the user as scale